### ISSUE TITLE
Bug: Surfshark should use hostnames instead of IP addresses

### ISSUE THREAD

**Issue by WoLpH** [qdm12/gluetun#244]
**TLDR**: *Surfshark uses a DNS based load balancer *

1. Is this urgent?

    - [ ] Yes
    - [x] No

2. What VPN service provider are you using?

    - [] PIA
    - [ ] Mullvad
    - [ ] Windscribe
    - [x] Surfshark
    - [ ] Cyberghost

3. What's the version of the program?

    **See the line at the top of your logs**

    Based on this file: https://github.com/qdm12/gluetun/blob/master/internal/constants/surfshark.go

4. What are you using to run the container?

    - [ ] Docker run
    - [] Docker Compose
    - [x] Kubernetes
    - [ ] Docker stack
    - [ ] Docker swarm
    - [ ] Podman
    - [ ] Other:

5. Extra information

Here's the result of a bunch of ` nslookup nl-ams.prod.surfshark.com`:

```
185.59.222.168
185.59.222.92
81.19.208.68
81.19.209.113
81.19.209.124
81.19.209.59
89.46.223.104
89.46.223.106
89.46.223.169
89.46.223.222
89.46.223.224
89.46.223.227
89.46.223.74
89.46.223.80
89.46.223.82
89.46.223.90
89.46.223.94
```

**Comment by qdm12** [comment]
Hi there, indeed. The code uses a `resolveRepeat` function to resolve multiple times each subdomain in order to get as many IPs as possible. I'll increase the repetition count (from 2/3 as I recall), is this what you are suggesting?

**Comment by WoLpH** [comment]
A quick test showed that they have 45 servers in Amsterdam. The current code has 2: https://github.com/qdm12/gluetun/blob/f9b6e854b10f14ef751a6cb5eca86ea4d94c13b8/internal/constants/surfshark.go#L96

I would suggest connecting to the hostname instead of the IP address and skip the IP lookups completely. Although I'm not sure how easy it would be to implement that in your current system.

**Comment by qdm12** [comment]
Doing so was how it was before, but it leaks DNS requests showing your ISP/government/whoever you are resolving a specific hostname for a certain region and vpn service.

You could use dns over TLS at start but I think some people have port 853 (dns over tls) or even dns servers IPs like 8.8.8.8 or 1.1.1.1 blocked.

I went through quite some coding ninja trouble to rely on IP addresses only. Note that in case the IPs are no longer valid, you can update the servers information using your network at your own risk or periodically using the env `UPDATE_PERIOD`. It's still undocumented, I need to add it to the wiki though.

But yes I'll increase the dns repetition count. I'm also working on making the resolutions faster (in a branch) as for now they are single threaded and synchronous, making many dns resolutions quite slow.

**Comment by WoLpH** [comment]
That makes sense. DoH could help you in that case but it's a valid concern indeed. I just noticed the issue and thought it was a bug. I'm not sure how much difference it makes in real life, but the 2 servers that are specified here could easily be overwhelmed if everyone starts using just those 2.

Too bad Surfshark doesn't appear to have an API to fetch the IP addresses.
I did notice this url which could be helpful for you: https://my.surfshark.com/vpn/api/v1/server/clusters

Oddly enough that API includes a load without including a specific server so I'm not sure what the load number means.

**Comment by qdm12** [comment]
I wrote a function to use Surfshark's clusters API endpoint you mentioned. Unfortunately, it seems it lacks several hostnames when compared with their zip file. Anyway I kept it in case we need a backup solution to fetch servers.

Also updating should now be faster as I parallelized the DNS resolutions. I increased the DNS repetition from 2-3 to 5 so more IP addresses should be in there.

I also updated the [Wiki page on updating the servers information yourself](https://github.com/qdm12/gluetun/wiki/Update-servers-information) if you feel like trying it.
